File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireHookStageRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to contain stage markers.
 */
public enum RequireHookStageRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireHookStage";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode stages = catNode.get("parameters").get("stages");
        final JsonNode stackStages = stages.get("maven");
        final String markerTemplate = catNode.get("parameters").get("markerTemplate").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return Stream.ofNullable(stackStages)
                .flatMap(ss -> Stream.of(ss.fieldNames().spliterator(), false)
                        .flatMap(hookName -> validateStage(root, hookName, ss.get(hookName).asText(), markerTemplate, severity).stream()))
                .toList();
    }

    private List<Finding> validateStage(Path root, String hookName, String expectedStage, String markerTemplate, String severity) throws MojoExecutionException {
        final String hookPath = "docs/harness/git-hooks/" + hookName;
        final Path hook = root.resolve(hookPath);
        return Stream.of(hook)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String text = HarnessCheckHelper.readFile(root, p);
                        final String expectedMarker = markerTemplate.replace("{stage}", expectedStage);
                        return text.contains(expectedMarker)
                                ? Stream.empty()
                                : Stream.of(new Finding(severity, CATEGORY, hookPath + " must contain stage marker '" + expectedMarker + "'"));
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireHookStageRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessCheckRule
import ai.harness.gradle.HarnessPsiResults
import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to contain stage markers.
 */
object RequireHookStageRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "requireHookStage"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "requireHookStage"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        val stagesObj = parametersObj?.get("stages")?.jsonObject
        val gradleStages = stagesObj?.get("gradle")?.jsonObject
        return if (catObj == null || parametersObj == null || messagesObj == null || stagesObj == null ||
            gradleStages == null
        ) {
            emptyList()
        } else {
            val markerTemplate = HarnessCheck.stringFrom(parametersObj, "markerTemplate")
            val preCommitStage = HarnessCheck.stringFrom(gradleStages, "pre-commit")
            val prePushStage = HarnessCheck.stringFrom(gradleStages, "pre-push")
            val preCommitHook = root / "docs/harness/git-hooks/pre-commit"
            val prePushHook = root / "docs/harness/git-hooks/pre-push"
            listOfNotNull(
                if (preCommitHook.isRegularFile()) {
                    val marker = markerTemplate.replace("{stage}", preCommitStage)
                    if (!preCommitHook.readText().contains(marker)) {
                        Finding(
                            HarnessCheck.severityOf(manifest, category),
                            category,
                            HarnessCheck.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() }
                                ?: "pre-commit must contain stage marker '$marker'",
                        )
                    } else {
                        null
                    }
                } else {
                    null
                },
                if (prePushHook.isRegularFile()) {
                    val marker = markerTemplate.replace("{stage}", prePushStage)
                    if (!prePushHook.readText().contains(marker)) {
                        Finding(
                            HarnessCheck.severityOf(manifest, category),
                            category,
                            HarnessCheck.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() }
                                ?: "pre-push must contain stage marker '$marker'",
                        )
                    } else {
                        null
                    }
                } else {
                    null
                },
            )
        }
    }
}
